src/main/java/com/traveltime/plugin/elasticsearch/FetcherSingleton.java

```diff
@@ -0,0 +1,19 @@
+package com.traveltime.plugin.elasticsearch;
+
+import java.net.URI;
+
+public enum FetcherSingleton {
+   INSTANCE;
+
+   private ProtoFetcher underlying = null;
+   private final Object lock = new Object[0];
+
+   public ProtoFetcher getFetcher(URI uri, String id, String key) {
+      if(underlying != null) return underlying;
+      synchronized (lock) {
+         if(underlying != null) return underlying;
```

Comment:
isn't this check redundant if there already was the same check before synchronization? Why not _only_ return `underlying` within `synchronized`?

Reply:
You check once before `synchronized`  because a contended lock is very expensive and unnecessary. The second check is because another thread might have assigned `underlying` before the current thread acquired the lock.

In our case dropping the second check doesn't hurt us much since we would at worst create a few SDK instances for the first request(s). But double-checked locking is a standard Java pattern and there's no reason not to do it.